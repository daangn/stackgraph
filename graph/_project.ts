import { Reducer, Stream } from "../deps/rimbu.ts"
import { Project, SourceFile } from "../deps/ts_morph.ts"
import { denoCompilerOptions } from "../utils/project.ts"

/** In-Memory ESM Project for quick testing */
export const inMemoryProject = () =>
	new Project({
		compilerOptions: denoCompilerOptions,
		useInMemoryFileSystem: true,
	})

/**
 * Add source files to in-memory project
 */
export const withSrc =
	(project: Project) =>
	<const T extends string>(src: Record<T, string>): Record<T, SourceFile> =>
		Stream.fromObject(src)
			.map(([path, text]) =>
				[path, project.createSourceFile(path, text)] as const
			)
			.reduce(Reducer.toJSObject())
