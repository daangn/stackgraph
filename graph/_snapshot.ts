const read = await Deno.permissions.query({ name: "read" })

const ignore = read.state !== "granted"

/**
 * same as `Deno.test` but skipped if `read` permission is denied
 */
export const snapshotTest = (
	name: Deno.TestDefinition["name"],
	fn: Deno.TestDefinition["fn"],
) => Deno.test({ name, ignore, fn })
