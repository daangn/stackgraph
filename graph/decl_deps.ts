import { Stream } from "../deps/rimbu.ts"
import {
	type ClassDeclaration,
	type FunctionDeclaration,
	type Node,
	ReferencedSymbol,
	ReferencedSymbolEntry,
	SyntaxKind,
	type VariableDeclaration,
} from "../deps/ts_morph.ts"
import { encodeVSCodeURI } from "./vscode_uri.ts"

export type Declaration =
	| VariableDeclaration
	| ClassDeclaration
	| FunctionDeclaration

export type DeclDeps = Map<Declaration, Declaration[]>

/**
 * Get the topmost declaration
 *
 * @privateRemarks
 *
 * FIXME(#1): handle destructuring and multiple variable declarations
 * RFC: only track top-level declarations or inside if statements?
 * FIXME: also track `function` statements, are there other types of declarations?
 */
export const getTopDecl = (
	ref: ReferencedSymbolEntry,
): Declaration | undefined => {
	const target = ref.getNode().getAncestors()

	// 	.at(-2) // HACK: last is SourceFile, second to last is the variable/class declaration

	// return (target?.isKind(SyntaxKind.VariableDeclaration) ||
	// 		target?.isKind(SyntaxKind.ClassDeclaration))
	// 	? target
	// 	: undefined

	return target.findLast((x): x is Declaration =>
		x.isKind(SyntaxKind.VariableDeclaration) ||
		x.isKind(SyntaxKind.ClassDeclaration) ||
		x.isKind(SyntaxKind.FunctionDeclaration)
	)
}

export const equals = (a: Node, b: Node) =>
	encodeVSCodeURI(a) === encodeVSCodeURI(b)

export const getActualReferences = (
	symbol: ReferencedSymbol,
): Stream<ReferencedSymbolEntry> =>
	Stream.from(symbol
		.getReferences())
		.filter((ref) => !equals(ref.getNode(), symbol.getDefinition().getNode()))
		.filter((ref) =>
			ref.getNode().getParent()?.getKindName() !== "ImportClause"
		)

export const getReferencedDecls = (
	node: Declaration,
): Stream<Declaration> =>
	Stream.from(node.findReferences())
		.flatMap(getActualReferences)
		.map((x) => getTopDecl(x))
		.filter((x): x is Declaration => x !== undefined)

export const getDeclDeps = (links: Declaration[]): DeclDeps => {
	const graph: DeclDeps = new Map()

	let current = links
	while (current.length > 0) {
		current = current.flatMap((node) => {
			const references = getReferencedDecls(node).toArray()
			graph.set(node, references)
			return references
		})
	}
	return graph
}
