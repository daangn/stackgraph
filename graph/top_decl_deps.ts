import { HashMap, HashSet } from "https://deno.land/x/rimbu@1.2.0/hashed/mod.ts"
import type { Graph } from "./graph.ts"
import { VSCodeURI } from "./vscode_uri.ts"

export type TopDeclDeps = HashMap<VSCodeURI, HashSet<VSCodeURI>>

/**
 * Flattens graph of references into top-level links
 *
 * ## Example
 *
 * ```
 * a -> A -> TOP1
 * b -> A
 * a -> B -> TOP2
 * c -> B
 * ```
 * will be flattened into
 * ```
 * TOP1 <- A, a, b
 * TOP2 <- B, a, c
 * ```
 */
export const graphToTopDeclDeps = (graph: Graph): TopDeclDeps => {
	const topLevelNodes = graph.streamNodes()
		.filter((node) => graph.getConnectionsFrom(node).size === 0)

	return topLevelNodes
		.map((page) =>
			[
				page,
				graph
					.getConnectionStreamFrom(page)
					.map(([, to]) => to)
					.filter((to) => to !== undefined)
					.reduce(HashSet.reducer()),
			] as const
		)
		.reduce(HashMap.reducer())
}
