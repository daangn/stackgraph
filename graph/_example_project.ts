import { Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { inMemoryProject } from "./_project.ts"

/**
 * https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements
 *
 * Example file of all top-level javascript declarations
 */
export const declExampleText = (h: string | number = "") => /*javascript*/ `
    // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements#declaring_variables
    export var varVar${h} = 1
    export var varVarIgnore${h} = 2

    export let varLet${h} = 1
    export let varLetIgnore${h} = 2

    export const varConst${h} = 1
    export const varConstIgnore${h} = 2

    export const varArrowFn${h} = () => {}
    export const varArrowFnIgnore${h} = () => {}

    // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements#functions_and_classes
    export function functionFn${h}() {}
    export function functionFnIgnore${h}() {}

    export function${h}* functionGen() {}
    export function${h}* functionGenIgnore() {}

    export async function${h} functionAsync() {}
    export async function${h} functionAsyncIgnore() {}

    export async function${h}* functionAsyncGen() {}
    export async function${h}* functionAsyncGenIgnore() {}

    export class Class${h} {}
    export class ClassIgnore${h} {}

    // others
    if (true) {
        const variableInIf${h} = 3
    }
`

/**
 * Arbitarily large file with many declarations for benchmarking
 */
export const getDeclExampleFile = (amount: number) => {
	const project = inMemoryProject()

	const text = Stream
		.range({ amount }).map(declExampleText).join({ sep: "\n\n" })

	return project.createSourceFile("a.ts", text)
}

export const exampleSrc = {
	"a.ts": /*javascript*/ `
        export const a = "a"

	    export const aaa = () => a
	`,
	"b.tsx": /*javascript*/ `
        import { a, aaa } from "./a.ts"
        export { a }

        export const Comp = () => <div>{a}</div>
        export const CompAAA = () => <div>{aaa()}</div>

        // unrelated to a and should be ignored
        export const Unrelated = () => <div>Unrelated</div>
    `,
	"c.tsx": /*javascript*/ `
        import { Comp } from "./b.tsx"

        export const Page = () => <div><Comp/></div>
    `,
	"d.tsx": /*javascript*/ `
        import { a } from "./b.tsx"
        import { Page } from "./c.tsx"

        export const InnerImport = () => <Page/>
        export const AliasedImport = () => <div>{a}</div>
    `,
}
