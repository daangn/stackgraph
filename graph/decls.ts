import {
	Stream,
	StreamSource,
} from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { SourceFile } from "../deps/ts_morph.ts"
import { Declaration } from "./decl_deps.ts"

/**
 * Find initial declarations to begin analyze with.
 */
export const getDecls =
	(filter: (node: Declaration) => boolean) =>
	(files: StreamSource<SourceFile>): Stream<Declaration> =>
		Stream.from(files).flatMap((file) =>
			Stream.from<Declaration>(
				file.getVariableDeclarations(),
				file.getFunctions(),
				file.getClasses(),
			).filter(filter)
		)

export const asMetadataRecord = <A, B>(links: LinkedDecl<A, B>[]) =>
	Object.fromEntries(
		links.map(({ node, ...metadata }) => [node.getName()!, metadata]),
	)
