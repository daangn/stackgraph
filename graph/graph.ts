import type { GraphElement } from "https://deno.land/x/rimbu@1.2.0/graph/custom/common/link.ts"
import { ArrowGraphHashed } from "https://deno.land/x/rimbu@1.2.0/graph/mod.ts"
import { Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { DeclDeps } from "./decl_deps.ts"

export type Id = string

/** directed graph of references, using `string` as unique key */
export type Graph = ArrowGraphHashed<Id>

/**
 * ## Warning
 *
 * uses `Node.getName` as unique key which is prone to collisions
 * if there are multiple variables with the same name
 *
 * FIXME: assumes class declaration has names (hence `!`)
 */
export const declDepsToGraph = (declDeps: DeclDeps): Graph =>
	Stream.from(declDeps).flatMap(([source, targets]) =>
		Stream.from(targets)
			.map((to): GraphElement<Id> => [source.getName()!, to.getName()!])
	)
		.reduce(ArrowGraphHashed.reducer())
