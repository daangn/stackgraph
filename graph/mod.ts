import { distinct } from "https://deno.land/std@0.215.0/collections/distinct.ts"
import {
	SourceFile,
	VariableDeclaration,
} from "https://deno.land/x/ts_morph@21.0.1/mod.ts"
import { getGraph } from "./graph.ts"
import type { Graph } from "./graph.ts"

export type Relations<A, B> = Record<string, Omit<LinkedVarDecl<A, B>, "decl">>
export type LinkedVarDecl<A, B> = {
	decl: VariableDeclaration
	links: A
	metas: B
}

export type Metadata = {
	logShowEvents: string[]
	logClickEvents: string[]
}

export type Dependencies = Record<string, string[]>

export type Rels = Record<string, Rel>
export type Rel = {
	links: Record<"push" | "replace", string[]>
	meta: Metadata
}

/**
 * creates a reference graph
 */
export const serialize = (graph: Graph) =>
	Array.from(graph.entries()).map(([decl, refs]) =>
		[decl.getName(), distinct(refs.map((x) => x.getName()))] as const
	)

type Option<A, B> = {
	/**
	 * Filter and map links, metadatas, and variable declaration
	 */
	getLink: (node: VariableDeclaration) => LinkedVarDecl<A, B> | undefined
}

export const generateGraph = <A, B>({ getLink }: Option<A, B>) =>
(
	files: SourceFile[],
): { dependencies: Dependencies; relations: Relations<A, B> } => {
	const links = files.flatMap((file) =>
		file.getVariableDeclarations().flatMap((decl) => getLink(decl) ?? [])
	)

	const relations = Object.fromEntries(
		links.map(({ decl, ...relations }) => [decl.getName(), relations]),
	)
	const dependencies = Object.fromEntries(
		links.flatMap(({ decl }) => serialize(getGraph(decl))),
	)

	return { dependencies, relations }
}
