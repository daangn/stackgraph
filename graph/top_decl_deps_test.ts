import { Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { assertSnapshot } from "../test_deps.ts"
import { exampleSrc } from "./_example_project.ts"
import { inMemoryProject, withSrc } from "./_project.ts"
import { getDeclDeps } from "./decl_deps.ts"
import { getAllDecls } from "./decls.ts"
import { declDepsToGraph } from "./graph.ts"
import { getTopDeclDeps } from "./top_decl_deps.ts"
import { snapshotTest } from "./_snapshot.ts"

snapshotTest(
	"getTopDeclDeps() converts graph into valid TopDeclDeps",
	async (t) => {
		const project = inMemoryProject()
		const files = withSrc(project)(exampleSrc)
		const decls = Stream.fromObjectValues(files).flatMap(getAllDecls).toArray()
		const declDeps = getDeclDeps(decls)
		const graph = declDepsToGraph(declDeps)

		const topDeclDeps = getTopDeclDeps(graph)

		await assertSnapshot(t, topDeclDeps)
	},
)
