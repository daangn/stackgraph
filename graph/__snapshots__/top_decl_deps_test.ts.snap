export const snapshot = {};

snapshot[`getTopDeclDeps() converts graph into valid TopDeclDeps 1`] = `
{
  "vscode://file/b.tsx:5:14?kind=VariableDeclaration&name=CompAAA": [
    "vscode://file/a.ts:3:14?kind=VariableDeclaration&name=aaa",
    "vscode://file/a.ts:1:14?kind=VariableDeclaration&name=a",
  ],
  "vscode://file/d.tsx:4:14?kind=VariableDeclaration&name=InnerImport": [
    "vscode://file/b.tsx:4:14?kind=VariableDeclaration&name=Comp",
    "vscode://file/a.ts:1:14?kind=VariableDeclaration&name=a",
    "vscode://file/c.tsx:7:14?kind=VariableDeclaration&name=Page",
    "vscode://file/b.tsx:8:14?kind=VariableDeclaration&name=Unrelated",
  ],
  "vscode://file/d.tsx:8:14?kind=VariableDeclaration&name=AliasedImport": [
    "vscode://file/a.ts:1:14?kind=VariableDeclaration&name=a",
  ],
}
`;
